README.md
# Redirectioner #

Creates automatic redirects for 404 traffic and page suggestions when matches are not found providing better service to your web visitors.

## Description ##

Redirectioner allows WordPress admins to have control over their dead links and redirects from inside the admin panel. Redirectioner records all URLs that users have visited and allows the admin to easily create 301 and 302 redirects to valid pages on their site. Redirects can also be created based on the best possible match for the URL the visitor was most likely trying to reach.

> <strong>Support & Bug Reports</strong><br>
> If you're in need of support or would like to file a bug report, please head over to our Github repository and [create a new issue.](https://github.com/ThemeMix/redirectioner/issues)
>

### Features: ###

* Get list of 404 URLs as they happen inside the admin panel
* Easily redirect 404 URLs to existing pages or choose to ignore the 404 error
* Provides the ability to automatically create redirects based on the URL the visitor was most likely trying to visit
* Provide visitors with a list of suggested pages on the 404 page when a automatic redirect can not be made
* Ability to suggest tag and category pages
* Ability to create automatic redirect for misspelled tag and category pages
* Ability to view logs of hits to 404 pages and redirects including referrer data
* Ability to remove automatically remove redirects when the URL matches a new page or post permalink
* Ability to automatically remove manual and automatic redirects once they are no longer being used
* All features work with both pages and posts
* Create automatic redirects for any URL resolving to a single page or post that isn't the current permalink
* Basic stats of plugin usage

Convert your 404 traffic by providing your site visitors with a better browsing experience and eliminate 404 URLs on your site.

## Installation ##

Installation of Redirectioner is simple:

1. Unzip `redirectioner.zip` and upload contents to the `/wp-content/plugins/` directory
1. Activate the plugin through the 'Plugins' menu in WordPress Admin
1. Use the `Settings -> Redirectioner` options page to enable desired features.

For the `suggested pages` feature to work you need to edit your 404.php template file to include the following code:
`<?php if (function_exists( 'wbz404_suggestions' ) ) { wbz404_suggestions(); } ?>`

## Frequently Asked Questions ##

### How long does it take for 404 URLs to start showing up? ###

As long as the "Capture incoming 404 URLs" option is enabled in the options section, the 404 URLs will show up in the captured list as soon as a visitor hits a 404 page.

### Will there be a slow down on my site when running the plugin? ###

No, there should be no noticeable slow down when running the plugin on your site.

### Will this plugin redirect my pages if I change my permalinks structure? ###

Yes! Redirectioner records the page/post ID number and looks up the most current permalink before redirecting the user.

### Can I redirect all 404's to a particular page? ###

No, that's not what this plugin is for. This plugin is designed to make your visitors experience better by automatically fixing 404 problems caused by typos.

## Screenshots ##

1. Admin Options Screen
![1. Admin Options Screen](https://ps.w.org/404-redirected/assets/screenshot-1.jpg)

2. Logs
![2. Logs](https://ps.w.org/404-redirected/assets/screenshot-2.jpg)

3. Create New Redirect
![3. Create New Redirect](https://ps.w.org/404-redirected/assets/screenshot-3.png)



## Upgrade Notice ##

## Version 1.4.9 ##
* Fixed parse error on PHP <5.4 due to use of short array syntax. [#30](https://github.com/ThemeMix/redirectioner/issues/30)

### Version 1.4.8 ###
* Rename to Redirectioner
* Update branding links
* Fixed broken pagination counts. [#17](https://github.com/ThemeMix/redirectioner/issues/17)
* Fixed PHP notices on admin screen. [#24](https://github.com/ThemeMix/redirectioner/issues/24)
* Fixed line-breaks in admin notices. [#25](https://github.com/ThemeMix/redirectioner/issues/25)
* Cleaned up query arguments handling. [#28](https://github.com/ThemeMix/redirectioner/issues/28)

### Version 1.4.7 ###
* Fixed too strict data sanitation for the `wbz404_suggestions()` template tag
* Fixed CSS class for suggested 404s div wrapper.

### Version 1.4.6 ###
* Fixed bug where query vars were being stripped
* Fixed a bug caused by plugin incorrectly injecting end-points turning up as 404s
* Fixed log purging issues
* General code improvements

### Version 1.4.4 ###
* Fixed a [SQL bug](https://github.com/ThemeMix/redirectioner/issues/7)
* Fixed a bug where [logs wouldn't get deleted](https://github.com/ThemeMix/redirectioner/issues/8)
* Fixed a bug where [deactivating and activating the plugin would reset the stats to 0](https://github.com/ThemeMix/redirectioner/issues/9)
* Fixed various [PHP notices](https://github.com/ThemeMix/redirectioner/issues/10)

### Version 1.3 ###

Added new purge options and sorting by number of hits. Lots of bug fixes.

### Version 1.2 ###

Major bug fixes. Also added bulk processing of URLs and admin notifications.

### Version 1.1 ###

2 bug fixes in adding/editing redirects

### Version 1.0 ###

Initial Release


CHANGELOG.md
# Changelog #

## Version 1.4.9 ##
* Fixed parse error on PHP <5.4 due to use of short array syntax. [#30](https://github.com/ThemeMix/redirectioner/issues/30)

## Version 1.4.8 ##
* Rename to Redirectioner
* Update branding links
* Fixed broken pagination counts. [#17](https://github.com/ThemeMix/redirectioner/issues/17)
* Fixed PHP notices on admin screen. [#24](https://github.com/ThemeMix/redirectioner/issues/24)
* Fixed line-breaks in admin notices. [#25](https://github.com/ThemeMix/redirectioner/issues/25)
* Cleaned up query arguments handling. [#28](https://github.com/ThemeMix/redirectioner/issues/28)

## Version 1.4.7 ##
* Fixed too strict data sanitation for the `wbz404_suggestions()` template tag
* Fixed CSS class for suggested 404s div wrapper.

## Version 1.4.6 ##
* Fixed bug where query vars were being stripped
* Fixed a bug caused by plugin incorrectly injecting end-points turning up as 404s
* Fixed log purging issues

## Version 1.4.4 ##
* Fixed a [SQL bug](https://github.com/defries/404-redirected/issues/7)
* Fixed a bug where [logs wouldn't get deleted](https://github.com/defries/404-redirected/issues/8)
* Fixed a bug where [deactivating and activating the plugin would reset the stats to 0](https://github.com/defries/404-redirected/issues/9)
* Fixed various [PHP notices](https://github.com/defries/404-redirected/issues/10)

## Version 1.4.3 ##

* Updating a bug where a check for an ancient MySQL version would throw an error
* Fixing integer bug in SQL query
* General debug errors fixed as well.

## Version 1.4.2 ##

* Introducing WordPress Coding Standards
* Replace `wpdb::escape` for `esc_sql()`
* Removing exotic translation function and replacing with default translation setup. In other words, the plugin is now translatable.

## Version 1.4.1 ##

Released: 2016-06-06

* Improved security hardening (bugfixing)

## Version 1.4.0 ##

* Plugin takeover from rrolfe because of lack of maintenance in four years
* Data sanitization added where needed (everywhere)
* Fixed PHP notices

## Version 1.3.2 ##

Released: 2012-08-29

New Features:

* None

Bug Fixes:

* Remember way back when 1.3.1 was released (yesterday) and I said the cron jobs were fixed? They weren't. This update fixes crons for users who are upgrading by implementing a version check on the DB.
* Added upgrade script functionality by implementing DB version checking
* Performed general code cleanup to get rid of PHP NOTICES

## Version 1.3.1 ##

Released: 2012-08-28

New Features:

* None

Bug Fixes:

* Fixed bug that caused cron jobs not to register properly - Stopped automatic deletion of redirects and logs from happening

## Version 1.3 ##

Released: 2012-08-28

New Features:

* Added ability to sort redirects by number of hits
* Added new tools tab to admin page with purge options

Bug Fixes:

* Fixed bug that caused plugin to ignore query string vars e.g. ?p=1234
* Fixed bug that caused duplicate redirects to be recorded
* Fixed bug causing feed URLs to be recorded as 404's
* Fixed bug causing preview URLs to be recorded as 404's
* Fixed bug causing trackback URLs to be recorded as 404's
* General Code Clean Up
* Fixed missing page title on Stats tab
* Fixed bug causing posts/pages using <!--nextpage--> tag to not allow visitor beyond page 1

## Version 1.2 ##

Released: 2011-12-06

New Features:

* Added cron to clean up duplicate redirects caused by simultaneous hits
* Added optional admin notifications when captured URLs goes over specified level
* Added new basic stats information from logs
* Modified default options to turn on automatic redirects for new installs
* Added bulk processing of captured and ignored URLs
* Added bulk deletion on trashed redirects and URLs

Bug Fixes:

* Fixed bug that caused plugin to remove and re-add redirects
* Fixed bug that caused logs to display nothing when ordered by URL

## Version 1.1 ##

Released: 2011-11-29

* Fixed adding and editing of Tag & Category redirects
* Fixed bug causing category redirects to incorrectly match external redirects

## Version 1.0 ##

Released: 2011-11-20

* Initial Release


404-redirected.php
<?php
/*
	Plugin Name: Redirectioner
	Plugin URI:  https://redirectioner.com
	Description: Creates automatic redirects for 404 traffic and page suggestions when matches are not found providing better service to your web visitors
	Author:      ThemeMix
	Author URI:  https://thememix.com

	Version: 1.4.10

	License:     GPL2
	License URI: https://www.gnu.org/licenses/gpl-2.0.html
	Domain Path: /languages
	Text Domain: 404-redirected

	Developed by Weberz Hosting. Maintained by ThemeMix.

	Copyright 2009  Weberz Hosting  (email: rob@weberz.com)
	Copyright 2016  ThemeMix

	This program is free software; you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation; either version 2 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

//Constants
define( 'WBZ404_URL', plugin_dir_url( __FILE__ ) );
define( 'WBZ404_PATH', plugin_dir_path( __FILE__ ) );
define( 'WBZ404_NAME', plugin_basename( __FILE__ ) );
define( 'WBZ404_VERSION', '1.4.9' );
define( 'WBZ404_HOME', 'https://redirectioner.com/' );
define( 'WBZ404_TRANS', 'wbz404_redirected' );

//URL Types
define( 'WBZ404_MANUAL', 1 );
define( 'WBZ404_AUTO', 2 );
define( 'WBZ404_CAPTURED', 3 );
define( 'WBZ404_IGNORED', 4 );

//Redirect Types
define( 'WBZ404_POST', 1 );
define( 'WBZ404_CAT', 2 );
define( 'WBZ404_TAG', 3 );
define( 'WBZ404_EXTERNAL', 4 );

require WBZ404_PATH."includes/functions.php";
require WBZ404_PATH."includes/frontend.php";

if ( is_admin() ) {
	require WBZ404_PATH."includes/admin.php";
}

/**
 * Load the text domain for translation of the plugin
 *
 * @since 1.4.2
 */
// Load text domain.
load_plugin_textdomain( '404-redirected', false, dirname( plugin_basename( __FILE__ ) ) . '/languages' );


includes/functions.php
<?php
/*
 * 404 Manager Global Functions
 *
*/

register_activation_hook( WBZ404_NAME, 'wbz404_pluginActivation' );
register_deactivation_hook( WBZ404_NAME, 'wbz404_pluginRemove' );

add_action( 'wbz404_duplicateCronAction', 'wbz404_removeDuplicatesCron' );
add_action( 'wbz404_cleanupCronAction', 'wbz404_cleaningCron' );


function wbz404_updateDBVersion() {
	$options = wbz404_getOptions( 1 );

	$options['DB_VERSION'] = WBZ404_VERSION;

	update_option( 'wbz404_settings', $options );

	return $options;
}

function wbz404_getOptions( $skip_db_check="0" ) {
	$options = get_option( 'wbz404_settings' );

	if ( $options == "" ) {
		add_option( 'wbz404_settings', '', '', 'no' );
	}

	// Check to make sure we aren't missing any new options.
	$defaults = wbz404_getDefaultOptions();
	$missing = false;
	foreach ( $defaults as $key => $value ) {
		if ( ! isset( $options[ $key ] ) || '' === $options[ $key ] ) {
			$options[ $key ] = $value;
			$missing = true;
		}
	}

	if ( $missing ) {
		update_option( 'wbz404_settings', $options );
	}

	if ( $skip_db_check == "0" ) {
		if ( $options['DB_VERSION'] != WBZ404_VERSION ) {
			if ( WBZ404_VERSION == "1.3.2" ) {
				//Unregister all crons. Some were bad.
				wbz404_unregisterCrons();

				//Register the good ones
				wbz404_registerCrons();
			}
			$options = wbz404_updateDBVersion();
		}
	}

	return $options;
}

function wbz404_getDefaultOptions() {
	$options = array(
		'default_redirect' => '301',
		'capture_404' => '1',
		'capture_deletion' => 1095,
		'manual_deletion' => '0',
		'admin_notification' => '200',
		'remove_matches' => '1',
		'display_suggest' => '1',
		'suggest_minscore' => '25',
		'suggest_max' => '5',
		'suggest_title' => '<h3>' . __( 'Suggested Alternatives', '404-redirected' ) . '</h3>',
		'suggest_before' => '<ol>',
		'suggest_after' => '</ol>',
		'suggest_entrybefore' => '<li>',
		'suggest_entryafter' => '</li>',
		'suggest_noresults' => '<p>' . __( 'No Results To Display.', '404-redirected' ) . '</p>',
		'suggest_cats' => '1',
		'suggest_tags' => '1',
		'auto_redirects' => '1',
		'auto_score' => '90',
		'auto_deletion' => '1095',
		'auto_cats' => '1',
		'auto_tags' => '1',
		'force_permalinks' => '1',
	);
	return $options;
}

function wbz404_pluginActivation() {
	global $wpdb;
	add_option( 'wbz404_settings', '', '', 'no' );

	$charset_collate = '';
	if ( !empty( $wpdb->charset ) ) {
		$charset_collate .= " DEFAULT CHARACTER SET $wpdb->charset";
	}
	if ( !empty( $wpdb->collate ) ) {
		$charset_collate .= " COLLATE $wpdb->collate";
	}
	$query = "CREATE TABLE IF NOT EXISTS `" . $wpdb->prefix . "wbz404_redirects` (
	  `id` bigint(30) NOT NULL auto_increment,
	  `url` varchar(512) NOT NULL,
	  `status` bigint(20) NOT NULL,
	  `type` bigint(20) NOT NULL,
	  `final_dest` varchar(512) NOT NULL,
	  `code` bigint(20) NOT NULL,
	  `disabled` int(10) NOT NULL default '0',
	  `timestamp` bigint(30) NOT NULL,
	  PRIMARY KEY  (`id`),
	  KEY `status` (`status`),
	  KEY `type` (`type`),
	  KEY `code` (`code`),
	  KEY `timestamp` (`timestamp`),
	  KEY `disabled` (`disabled`),
	  FULLTEXT KEY `url` (`url`),
	  FULLTEXT KEY `final_dest` (`final_dest`)
	) ENGINE=MyISAM " . esc_html( $charset_collate ) . " COMMENT='404 Redirected Plugin Redirects Table' AUTO_INCREMENT=1";
	$wpdb->query( $query );

	$query = "CREATE TABLE IF NOT EXISTS `" . $wpdb->prefix . "wbz404_logs` (
	  `id` bigint(40) NOT NULL auto_increment,
	  `redirect_id` bigint(40) NOT NULL,
	  `timestamp` bigint(40) NOT NULL,
	  `remote_host` varchar(512) NOT NULL,
	  `referrer` varchar(512) NOT NULL,
	  `action` varchar(512) NOT NULL,
	  PRIMARY KEY  (`id`),
	  KEY `redirect_id` (`redirect_id`),
	  KEY `timestamp` (`timestamp`)
	) ENGINE=MyISAM " . esc_html( $charset_collate ) . " COMMENT='404 Redirected Plugin Logs Table' AUTO_INCREMENT=1";
	$wpdb->query( $query );

	wbz404_registerCrons();

	$options = wbz404_updateDBVersion();
}

function wbz404_registerCrons() {
	$timestamp = wp_next_scheduled( 'wbz404_cleanupCronAction' );
	if ( $timestamp == False ) {
		wp_schedule_event( current_time( 'timestamp' ) - 86400, 'daily', 'wbz404_cleanupCronAction' );
	}

	$timestamp = wp_next_scheduled( 'wbz404_duplicateCronAction' );
	if ( $timestamp == False ) {
		wp_schedule_event( current_time( 'timestamp' ) - 3600, 'hourly', 'wbz404_duplicateCronAction' );
	}
}

function wbz404_unregisterCrons() {

	$crons = array( 'wbz404_cleanupCronAction', 'wbz404_duplicateCronAction', 'wbz404_removeDuplicatesCron', 'wbz404_cleaningCron' );
	for ( $i=0; $i < count( $crons ); $i++ ) {
		$cron_name = $crons[$i];
		$timestamp = wp_next_scheduled( $cron_name );
		while ( $timestamp != False ) {
			wp_unschedule_event( $timestamp, $cron_name );
			$timestamp = wp_next_scheduled( $cron_name );
		}

		$timestamp = wp_next_scheduled( $cron_name, '' );
		while ( $timestamp != False ) {
			wp_unschedule_event( $timestamp, $cron_name, '' );
			$timestamp = wp_next_scheduled( $cron_name, '' );
		}

		wp_clear_scheduled_hook( $cron_name );
	}
}

function wbz404_pluginRemove() {
	//delete_option( 'wbz404_settings' );
	wbz404_unregisterCrons();
}

function wbz404_rankPermalinks( $url, $includeCats = '1', $includeTags = '1' ) {
	global $wpdb;
	$permalinks = array();

	$query = "select id from $wpdb->posts where post_status='publish' and (post_type='page' or post_type='post')";
	$rows = $wpdb->get_results( $query );
	foreach ( $rows as $row ) {
		$id = $row->id;
		$the_permalink = get_permalink( $id );
		$urlParts = parse_url( $the_permalink );
		$scoreBasis = strlen( $urlParts['path'] );
		$levscore = levenshtein( $url, $urlParts['path'], 1, 1, 1 );
		$score = 100 - ( ( $levscore / $scoreBasis )*100 );
		$permalinks[$id . "|POST"] = number_format( $score, 4, '.', '' );
	}

	if ( $includeTags == "1" ) {
		$query = "select " . esc_html( $wpdb->terms ) . ".term_id from " . esc_html( $wpdb->terms ) . " ";
		$query .= "left outer join " . esc_html( $wpdb->term_taxonomy ) . " on " . esc_html( $wpdb->terms ) . ".term_id = " . esc_html( $wpdb->term_taxonomy ) . ".term_id ";
		$query .= "where " . esc_html( $wpdb->term_taxonomy ) . ".taxonomy='post_tag' and " . esc_html( $wpdb->term_taxonomy ) . ".count >= 1";
		$rows = $wpdb->get_results( $query );
		foreach ( $rows as $row ) {
			$id = $row->term_id;
			$the_permalink = get_tag_link( $id );
			$urlParts = parse_url( $the_permalink );
			$scoreBasis = strlen( $urlParts['path'] );
			$levscore = levenshtein( $url, $urlParts['path'], 1, 1, 1 );
			$score = 100 - ( ( $levscore / $scoreBasis )*100 );
			$permalinks[$id . "|TAG"] = number_format( $score, 4, '.', '' );
		}
	}

	if ( $includeCats == "1" ) {
		$query = "select " . esc_html( $wpdb->terms ) . ".term_id from " . esc_html( $wpdb->terms ) . " ";
		$query .= "left outer join " . esc_html( $wpdb->term_taxonomy ) . " on " . esc_html( $wpdb->terms ) . ".term_id = " . esc_html( $wpdb->term_taxonomy ) . ".term_id ";
		$query .= "where " . esc_html( $wpdb->term_taxonomy ) . ".taxonomy='category' and " . esc_html( $wpdb->term_taxonomy ) . ".count >= 1";
		$rows = $wpdb->get_results( $query );
		foreach ( $rows as $row ) {
			$id = $row->term_id;
			$the_permalink = get_category_link( $id );
			$urlParts = parse_url( $the_permalink );
			$scoreBasis = strlen( $urlParts['path'] );
			$levscore = levenshtein( $url, $urlParts['path'], 1, 1, 1 );
			$score = 100 - ( ( $levscore / $scoreBasis )*100 );
			$permalinks[$id . "|CAT"] = number_format( $score, 4, '.', '' );
		}
	}

	arsort( $permalinks );
	return $permalinks;
}

function wbz404_permalinkInfo( $k, $v ) {
	$permalink = array();

	$meta = $k;
	$meta = explode( "|", $meta );

	$permalink['id'] = $meta[0];
	$permalink['type'] = $meta[1];
	$permalink['score'] = $v;

	if ( $permalink['type'] == "POST" ) {
		$permalink['link'] = get_permalink( $permalink['id'] );
		$permalink['title'] = get_the_title( $permalink['id'] );
	} else if ( $permalink['type'] == "TAG" ) {
			$permalink['link'] = get_tag_link( $permalink['id'] );
			$tag = get_term( $permalink['id'], 'post_tag' );
			$permalink['title'] = $tag->name;
		} else if ( $permalink['type'] == "CAT" ) {
			$permalink['link'] = get_category_link( $permalink['id'] );
			$cat = get_term( $permalink['id'], 'category' );
			$permalink['title'] = $cat->name;
		}

	return $permalink;
}


function wbz404_loadRedirectData( $url ) {
	global $wpdb;
	$redirect = array();

	$query="select * from " . esc_html( $wpdb->prefix ) . "wbz404_redirects where url = '" . esc_sql( esc_url( $url ) ) . "'";

	$row = $wpdb->get_row( $query, ARRAY_A );
	if ( $row == NULL ) {
		$redirect['id']=0;
	} else {
		$redirect['id'] = $row['id'];
		$redirect['url'] = $row['url'];
		$redirect['status'] = $row['status'];
		$redirect['type'] = $row['type'];
		$redirect['final_dest'] = $row['final_dest'];
		$redirect['code'] = $row['code'];
		$redirect['disabled'] = $row['disabled'];
		$redirect['created'] = $row['timestamp'];
	}
	return $redirect;
}

function wbz404_setupRedirect( $url, $status, $type, $final_dest, $code, $disabled = 0 ) {
	global $wpdb;

	$now = time();
	$wpdb->insert( $wpdb->prefix . 'wbz404_redirects',
		array(
			'url' => esc_url( $url ),
			'status' => esc_html( $status ),
			'type' => esc_html( $type ),
			'final_dest' => esc_html( $final_dest ),
			'code' => esc_html( $code ),
			'disabled' => esc_html( $disabled ),
			'timestamp' => esc_html( $now )
		),
		array(
			'%s',
			'%d',
			'%d',
			'%s',
			'%d',
			'%d',
			'%d'
		)
	);
	return $wpdb->insert_id;
}

function wbz404_logRedirectHit( $id, $action ) {
	global $wpdb;
	$now = time();

	if ( isset( $_SERVER['HTTP_REFERER'] ) ) {
		$referer=$_SERVER['HTTP_REFERER'];
	} else {
		$referer = "";
	}

	$wpdb->insert( $wpdb->prefix . "wbz404_logs",
		array(
			'redirect_id' => absint( $id ),
			'timestamp' => esc_html( $now ),
			'remote_host' => esc_html( $_SERVER['REMOTE_ADDR'] ),
			'referrer' => esc_html( $referer ),
			'action' => esc_html( $action ),
		),
		array(
			'%d',
			'%d',
			'%s',
			'%s',
			'%s'
		)
	);
}

function wbz404_cleanRedirect($id) {
	global $wpdb;
	if ($id != "" && $id != '0') {
		$id = absint( $id );
		$id = (string) $id;
		$query="delete from " . $wpdb->prefix . "wbz404_redirects where id = " . $wpdb->escape($id);
		$wpdb->query($query);
		$query="delete from " . $wpdb->prefix . "wbz404_logs where redirect_id = " . $wpdb->escape($id);
		$wpdb->query($query);
	}
}

function wbz404_cleaningCron() {
	global $wpdb;
	$options = wbz404_getOptions();
	$now = time();

	//Remove Captured URLs
	if ( $options['capture_deletion'] != '0' ) {
		$capture_time = $options['capture_deletion'] * 86400;
		$then = $now - $capture_time;

		//Clean up old logs
		$query = "delete from " . $wpdb->prefix . "wbz404_logs where ";
		$query .= "redirect_id in (select id from " . $wpdb->prefix . "wbz404_redirects where status = " . esc_sql( WBZ404_CAPTURED ) . " or status = " . esc_sql( WBZ404_IGNORED ) . ") ";
		$query .= "and timestamp < " . esc_sql( $then );
		$wpdb->query( $query );

		//Find unused urls
		$query = "select id from " . $wpdb->prefix . "wbz404_redirects where (status = " . esc_sql( WBZ404_CAPTURED ) . " or status = " . esc_sql( WBZ404_IGNORED ) . ") and ";
		$query .= "timestamp <= " . esc_sql( $then ) . " and id not in (";
		$query .= "select redirect_id from " . $wpdb->prefix . "wbz404_logs";
		$query .= ")";
		$rows = $wpdb->get_results( $query, ARRAY_A );
		foreach ( $rows as $row ) {
			//Remove Them
			wbz404_cleanRedirect( $row['id'] );
		}
	}

	//Remove Automatic Redirects
	if ( $options['auto_deletion'] != '0' ) {
		$auto_time = $options['auto_deletion'] * 86400;
		$then = $now - $auto_time;

		//Clean up old logs
		$query = "delete from " . $wpdb->prefix . "wbz404_logs where ";
		$query .= "redirect_id in (select id from " . $wpdb->prefix . "wbz404_redirects where status = " . esc_sql( WBZ404_AUTO ) . ") ";
		$query .= "and timestamp < " . esc_sql( $then );
		$wpdb->query( $query );

		//Find unused urls
		$query = "select id from " . $wpdb->prefix . "wbz404_redirects where status = " . esc_sql( WBZ404_AUTO ) . " and ";
		$query .= "timestamp <= " . esc_sql( $then ) . " and id not in (";
		$query .= "select redirect_id from " . $wpdb->prefix . "wbz404_logs";
		$query .= ")";
		$rows = $wpdb->get_results( $query, ARRAY_A );
		foreach ( $rows as $row ) {
			//Remove Them
			wbz404_cleanRedirect( $row['id'] );
		}
	}

	//Remove Manual Redirects
	if ( $options['manual_deletion'] != '0' ) {
		$manual_time = $options['manual_deletion'] * 86400;
		$then = $now - $manual_time;

		//Clean up old logs
		$query = "delete from " . $wpdb->prefix . "wbz404_logs where ";
		$query .= "redirect_id in (select id from " . $wpdb->prefix . "wbz404_redirects where status = " . esc_sql( WBZ404_MANUAL ) . ") ";
		$query .= "and timestamp < " . esc_sql( $then );
		$wpdb->query( $query );

		//Find unused urls
		$query = "select id from " . $wpdb->prefix . "wbz404_redirects where status = " . esc_sql( WBZ404_MANUAL ) . " and ";
		$query .= "timestamp <= " . esc_sql( $then ) . " and id not in (";
		$query .= "select redirect_id from " . $wpdb->prefix . "wbz404_logs";
		$query .= ")";
		$rows = $wpdb->get_results( $query, ARRAY_A );
		foreach ( $rows as $row ) {
			//Remove Them
			wbz404_cleanRedirect( $row['id'] );
		}
	}
}

function wbz404_removeDuplicatesCron() {
	global $wpdb;

	$rtable = $wpdb->prefix . "wbz404_redirects";
	$ltable = $wpdb->prefix . "wbz404_logs";

	$query = "SELECT COUNT(id) as repetitions, url FROM " . esc_html( $rtable ) . " GROUP BY url HAVING repetitions > 1";
	$rows = $wpdb->get_results( $query, ARRAY_A );
	foreach ( $rows as $row ) {
		$url = $row['url'];

		$query2 = "select id from " . esc_html( $rtable ) . " where url = '" . esc_sql( esc_url( $url ) ) . "' order by id limit 0,1";
		$orig = $wpdb->get_row( $query2, ARRAY_A, 0 );
		if ( $orig['id'] != 0 ) {
			$original = $orig['id'];

			//Fix the logs table
			$query2 = "update " . $ltable . " set redirect_id = " . esc_sql( $original ) . " where redirect_id in (select id from " . esc_html( $rtable ) . " where url = '" . esc_sql( $url ) . "' and id != " . esc_sql( $original ) . ")";
			$wpdb->query( $query2 );

			$query2 = "delete from " . $rtable . " where url='" . esc_sql( esc_url( $url ) ) . "' and id != " . esc_sql( $original );
			$wpdb->query( $query2 );

		}
	}

}

function wbz404_SortQuery( $urlParts ) {
	$url = "";
	if ( isset( $urlParts['query'] ) && $urlParts['query'] != "" ) {
		$queryString = array();
		$urlQuery = $urlParts['query'];
		$queryParts = preg_split( "/[;&]/", $urlQuery );
		foreach ( $queryParts as $query ) {
			if ( strpos( $query, "=" ) === false ) {
				$queryString[$query]='';
			} else {
				$stringParts = preg_split( "/=/", $query );
				$queryString[$stringParts[0]]=$stringParts[1];
			}
		}
		ksort( $queryString );
		$x=0;
		$newQS = "";
		foreach ( $queryString as $key => $value ) {
			if ( $x != 0 ) {
				$newQS .= "&";
			}
			$newQS .= $key;
			if ( $value != "" ) {
				$newQS .= "=" . $value;
			}
			$x++;
		}

		if ( $newQS != "" ) {
			$url .= "?" . $newQS;
		}
	}
	return esc_url( $url );
}

function wbz404_ProcessRedirect( $redirect ) {
	//A redirect record has already been found.
	if ( ( $redirect['status'] == WBZ404_MANUAL || $redirect['status'] == WBZ404_AUTO ) && $redirect['disabled'] == 0 ) {
		//It's a redirect, not a captured or ignored URL
		if ( $redirect['type'] == WBZ404_EXTERNAL ) {
			//It's a external url setup by the user
			wbz404_logRedirectHit( $redirect['id'], $redirect['final_dest'] );
			wp_redirect( esc_url( $redirect['final_dest'] ), esc_html( $redirect['code'] ) );
			exit;
		} else {
			$key="";
			if ( $redirect['type'] == WBZ404_POST ) {
				$key = $redirect['final_dest'] . "|POST";
			} else if ( $redirect['type'] == WBZ404_CAT ) {
				$key = $redirect['final_dest'] . "|CAT";
			} else if ( $redirect['type'] == WBZ404_TAG ) {
				$key = $redirect['final_dest'] . "|TAG";
			}
			if ( $key != "" ) {
				$permalink = wbz404_permalinkInfo( $key, 0 );
				wbz404_logRedirectHit( $redirect['id'], $permalink['link'] );
				wp_redirect( esc_url( $permalink['link'] ), esc_html( $redirect['code'] ) );
				exit;
			}
		}
	} else {
		wbz404_logRedirectHit( esc_html( $redirect['id'] ), '404' );
	}
}
